src/bvals/boundary_conditions_apk.hpp

```diff
@@ -0,0 +1,90 @@
+// AthenaPK - a performance portable block structured AMR astrophysical MHD code.
+// Copyright (c) 2025, Athena-Parthenon Collaboration. All rights reserved.
+// Licensed under the 3-clause BSD License, see LICENSE file for details
+//========================================================================================
+//! \file boundary_conditions_apk.chpp
+//  \brief AthenaPK specific boundary conditions
+//
+
+#ifndef BVALS_BOUNDARY_CONDITIONS_APK_HPP_
+#define BVALS_BOUNDARY_CONDITIONS_APK_HPP_
+
+#include <memory>
+#include <string>
+#include <vector>
+
+// Parthenon headers
+#include <parthenon/package.hpp>
+
+#include "basic_types.hpp"
+#include "bvals/boundary_conditions_generic.hpp"
+#include "mesh/domain.hpp"
+#include "mesh/mesh.hpp"
+#include "mesh/meshblock.hpp"
+#include "utils/error_checking.hpp"
+
+#include "../main.hpp"
+
+namespace Hydro {
+namespace BoundaryFunction {
+
+using namespace parthenon::package::prelude;
+using parthenon::CoordinateDirection;
+// using parthenon::MeshBlockData;
+// using parthenon::Real;
+using parthenon::BoundaryFunction::BCSide;
+
+template <CoordinateDirection DIR, BCSide SIDE>
+void ReflectBC(std::shared_ptr<MeshBlockData<Real>> &mbd, bool coarse) {
+  // make sure DIR is X[123]DIR so we don't have to check again
+  static_assert(DIR == X1DIR || DIR == X2DIR || DIR == X3DIR, "DIR must be X[123]DIR");
+
+  MeshBlock *pmb = mbd->GetBlockPointer();
+
+  auto hydro_pkg = pmb->packages.Get("Hydro");
+  auto fluid = hydro_pkg->Param<Fluid>("fluid");
+  PARTHENON_REQUIRE_THROWS(
+      fluid == Fluid::euler,
+      "Reflecting boundary conditions for MHD need special treatment.");
+
+  // convenient shorthands
+  constexpr bool X1 = (DIR == X1DIR);
+  constexpr bool X2 = (DIR == X2DIR);
+  constexpr bool X3 = (DIR == X3DIR);
+  constexpr bool INNER = (SIDE == BCSide::Inner);
+
+  const auto &bounds = coarse ? pmb->c_cellbounds : pmb->cellbounds;
+
+  const auto &range = X1 ? bounds.GetBoundsI(IndexDomain::interior)
+                         : (X2 ? bounds.GetBoundsJ(IndexDomain::interior)
+                               : bounds.GetBoundsK(IndexDomain::interior));
+  const int ref = INNER ? range.s : range.e;
+
+  constexpr IndexDomain domain =
+      INNER ? (X1 ? IndexDomain::inner_x1
+                  : (X2 ? IndexDomain::inner_x2 : IndexDomain::inner_x3))
+            : (X1 ? IndexDomain::outer_x1
+                  : (X2 ? IndexDomain::outer_x2 : IndexDomain::outer_x3));
+
+  // used for reflections
+  const int offset = (2 * ref) + (INNER ? -1 : 1);
+
+  auto cons = mbd->PackVariables(std::vector<std::string>{"cons"}, coarse);
+  const bool fine = false; // no usage of fine fields in AthenaPK for now
+
+  const auto nv = IndexRange{0, cons.GetDim(4) - 1};
+  pmb->par_for_bndry(
+      "ReflectBC", nv, domain, parthenon::TopologicalElement::CC, coarse, fine,
+      KOKKOS_LAMBDA(const int &v, const int &k, const int &j, const int &i) {
+        const bool reflect = v == DIR;
```

Comment:
Why does this enum / int comparison work? If the order of the hydro components in AthenaPK were to change, would this still work?

Reply:
Correct, it only works because of the current (fixed) ordering.
I don't see an issue right now (as no fundamental change of this is planned until the boundary comm in Parthenon is more efficient for many small variables), but to be sure that we don't miss this in the future, I added a regression test, which should fail if sth along those lines changes.